### PR TITLE
Task 6 7

### PR DESCRIPTION
**Pull request by nuriddin2468**

### CODE REVIEW

File: src/app/modules/courses/components/course-add-edit/course-add-edit.component.spec.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    })
+    .compileComponents();
+
+    fixture = TestBed.createComponent(CourseAddEditComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
```
Comment (goremchik) [inline src/app/modules/courses/components/course-add-edit/course-add-edit.component.spec.ts:43]: Add a few tests to your new component

File: src/app/shared/components/wrapper/header/header.component.ts [hunk 2]
```diff
@@ -1,15 +1,24 @@
...
 export class HeaderComponent implements OnInit {
-
-  constructor() { }
+  user: User;
+  constructor(
+    private authService: AuthService
+  ) { }
 
   ngOnInit(): void {
+    this.authService.getUserInfo().subscribe(user => this.user = user);
```
Comment (goremchik) [inline src/app/shared/components/wrapper/header/header.component.ts]: unsubscribe from observable

File: src/app/shared/guards/auth.guard.spec.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+describe('AuthGuard', () => {
+  let guard: AuthGuard;
+
+  beforeEach(() => {
+    TestBed.configureTestingModule({});
+    guard = TestBed.inject(AuthGuard);
+  });
+
+  it('should be created', () => {
+    expect(guard).toBeTruthy();
```
Comment (goremchik) [inline src/app/shared/guards/auth.guard.spec.ts:14]: Add tests

File: src/app/shared/guards/auth.guard.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  constructor(
+    private authService: AuthService,
+    private router: Router
+  ) {
+  }
+  canActivateChild(
+    route: ActivatedRouteSnapshot,
+    state: RouterStateSnapshot): Observable<boolean | UrlTree> | Promise<boolean | UrlTree> | boolean | UrlTree {
+    return this.authService.isAuthenticated()
+      .pipe(tap(res => {
```
Comment (goremchik) [inline src/app/shared/guards/auth.guard.ts]: No need to use tap here. You can use map return UrlTree, which redirects

File: src/app/modules/courses/components/course-add-edit/course-add-edit.component.html [hunk 5]
```diff
@@ -0,0 +1,88 @@
+<div class="card">
+  <form [formGroup]="form" (submit)="save()">
+    <h2 *ngIf="isNew else currentCourse">New course</h2>
```
Comment (goremchik) [inline src/app/modules/courses/components/course-add-edit/course-add-edit.component.html]: <h2>{{isNew  ? 'New course' : (form.value['title'] | uppercase )}}</h2>

File: src/app/modules/courses/components/course-add-edit/course-add-edit.component.html [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    <ng-template #currentCourse>
+      <h2>{{form.value['title'] | uppercase}}</h2>
+    </ng-template>
+    <div class="input-field">
```
Comment (goremchik) [inline src/app/modules/courses/components/course-add-edit/course-add-edit.component.html]: It is too much duplicates related to this block. Maybe you can move as separate component some part and reuse them

File: src/app/modules/courses/components/course-add-edit/course-add-edit.component.html (hunk 5 above)
Comment (nuriddin2468) [reply src/app/modules/courses/components/course-add-edit/course-add-edit.component.html]: fixed

File: src/app/shared/guards/auth.guard.ts (hunk 4 above)
Comment (nuriddin2468) [reply src/app/shared/guards/auth.guard.ts]: fixed

File: src/app/shared/components/wrapper/header/header.component.ts (hunk 2 above)
Comment (nuriddin2468) [reply src/app/shared/components/wrapper/header/header.component.ts]: fixed

File: src/app/modules/courses/components/course-add-edit/course-add-edit.component.spec.ts (hunk 1 above)
Comment (nuriddin2468) [reply src/app/modules/courses/components/course-add-edit/course-add-edit.component.spec.ts:43]: done

File: src/app/shared/guards/auth.guard.spec.ts (hunk 3 above)
Comment (nuriddin2468) [reply src/app/shared/guards/auth.guard.spec.ts:14]: cannot write :(

File: src/app/modules/courses/components/course-add-edit/course-add-edit.component.html (hunk 6 above)
Comment (nuriddin2468) [reply src/app/modules/courses/components/course-add-edit/course-add-edit.component.html]: done

File: src/app/shared/guards/auth.guard.spec.ts (hunk 3 above)
Comment (goremchik) [reply src/app/shared/guards/auth.guard.spec.ts:14]: Why? You don't need real service in order to test guard

File: src/app/shared/components/breadcrumbs/breadcrumbs.component.ts [hunk 7]
```diff
@@ -7,9 +9,33 @@ import { Component, OnInit } from '@angular/core';
...
+    private route: ActivatedRoute,
+    private router: Router
+  ) {
   }
 
+  ngOnInit() {
+    this.createBreadcrumbs();
+    this.router.events.pipe(
+      filter(event => event instanceof NavigationEnd)
+    ).subscribe(() => {
```
Comment (goremchik) [inline src/app/shared/components/breadcrumbs/breadcrumbs.component.ts:27]: Remove subscription on destroy

File: src/app/shared/components/breadcrumbs/breadcrumbs.component.html [hunk 8]
```diff
@@ -1,3 +1,7 @@
 <div class="content">
-  Courses
+  <ul>
+    <li *ngFor="let item of breadcrumbs; last as last">
+      {{item}} <span class="slash">{{last ? '' : '/'}}</span>
```
Comment (goremchik) [inline src/app/shared/components/breadcrumbs/breadcrumbs.component.html]: This slash could be added with styles only

Comment (nuriddin2468) [reply src/app/shared/components/breadcrumbs/breadcrumbs.component.html]: done

File: src/app/shared/components/breadcrumbs/breadcrumbs.component.ts (hunk 7 above)
Comment (nuriddin2468) [reply src/app/shared/components/breadcrumbs/breadcrumbs.component.ts:27]: done

**Review by goremchik — APPROVED** [review]

**Pull request merged**